common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/PasskeyOriginRulesManager.kt

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.providers.oauth2
+
+import com.microsoft.identity.common.logging.Logger
+import java.net.URI
+
+/**
+ * Manages allowed origin rules for Passkey/WebAuthN APIs.
+ * Ensures only trusted Microsoft origins can access sensitive credential operations.
+ *
+ * **Origin Categories:**
+ * 1. **Production origins** - Allow all paths (any path is safe)
+ * 2. **Sovereign cloud origins** - Require "/fido" as a path segment
+ *
+ * **Security Features:**
+ * - Validates scheme must be HTTPS
+ * - Prevents subdomain spoofing (e.g., attacker.com.microsoft.com)
+ * - Prevents prefix attacks (e.g., evillogin.microsoft.com)
+ * - Requires exact domain matching
+ * - Case-insensitive FIDO path matching for sovereign clouds
+ */
+object PasskeyOriginRulesManager {
+    private const val TAG = "PasskeyOriginRulesManager"
+    private const val HTTPS_SCHEME = "https"
+    private const val FIDO_SEGMENT = "fido"
+
+    // Production origins - allow any path
+    private val PRODUCTION_ORIGINS = setOf(
```

Comment:
nit: if pre-prod origins are getting removed (btw, is there a particular reason why pre-prod support is getting removed?), maybe the name should be something like public_cloud_origins or similar